File: EPI/14.8/swift/14.8_test.swift
import Darwin

func == (lhs: [Student], rhs: [Student]) -> Bool {
    guard lhs.count == rhs.count else {
        return false
    }
    for (i, v) in lhs.enumerated() {
        guard v == rhs[i] else {
            return false
        }
    }
    return true
}

let tests: [(students: [Student], grouped: [Student])] = [
    (
        students: [],
        grouped: []
    ),
    (
        students: [
            Student(name: "17 #1", age: 17),
            Student(name: "18 #1", age: 18),
        ],
        grouped: [
            Student(name: "17 #1", age: 17),
            Student(name: "18 #1", age: 18),
        ]
    ),
    (
        students: [
            Student(name: "18 #1", age: 18),
            Student(name: "17 #1", age: 17),
        ],
        grouped: [
            Student(name: "17 #1", age: 17),
            Student(name: "18 #1", age: 18),
        ]
    ),
    (
        students: [
            Student(name: "17 #1", age: 17),
            Student(name: "18 #1", age: 18),
            Student(name: "17 #2", age: 17),
        ],
        grouped: [
            Student(name: "17 #1", age: 17),
            Student(name: "17 #2", age: 17),
            Student(name: "18 #1", age: 18),
        ]
    ),
    (
        students: [
            Student(name: "17 #1", age: 17),
            Student(name: "18 #1", age: 18),
            Student(name: "21 #1", age: 21),
            Student(name: "21 #2", age: 21),
            Student(name: "18 #2", age: 18),
            Student(name: "17 #2", age: 17),
            Student(name: "21 #3", age: 21),
            Student(name: "19 #1", age: 19),
        ],
        grouped: [
            Student(name: "17 #1", age: 17),
            Student(name: "17 #2", age: 17),
            Student(name: "18 #1", age: 18),
            Student(name: "18 #2", age: 18),
            Student(name: "19 #1", age: 19),
            Student(name: "21 #1", age: 21),
            Student(name: "21 #2", age: 21),
            Student(name: "21 #3", age: 21),
        ]
    ),
]

for test in tests {
    var students = test.students
    group(&students)
    guard students == test.grouped else {
        print("For test students \(test.students), " +
            "expected grouped students to be \(test.grouped), but were \(students)")
        exit(1)
    }
}

File: EPI/Sorting/Partitioning and sorting an array with many repeated entries/typescript/14.8_test.ts
import {Student, rearrangeInPlace} from './14.8';

declare function require(name: string): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

interface TestStudent extends Student {
    name: string;
}

const tests: {students: TestStudent[], rearranged: TestStudent[]}[] = [
    {
        students: [],
        rearranged: []
    },
    {
        students: [
            {name: '17 #1', age: 17},
            {name: '18 #1', age: 18},
        ],
        rearranged: [
            {name: '17 #1', age: 17},
            {name: '18 #1', age: 18},
        ]
    },
    {
        students: [
            {name: '18 #1', age: 18},
            {name: '17 #1', age: 17},
        ],
        rearranged: [
            {name: '17 #1', age: 17},
            {name: '18 #1', age: 18},
        ]
    },
    {
        students: [
            {name: '17 #1', age: 17},
            {name: '18 #1', age: 18},
            {name: '17 #2', age: 17},
        ],
        rearranged: [
            {name: '17 #1', age: 17},
            {name: '17 #2', age: 17},
            {name: '18 #1', age: 18},
        ]
    },
    {
        students: [
            {name: '17 #1', age: 17},
            {name: '18 #1', age: 18},
            {name: '21 #1', age: 21},
            {name: '21 #2', age: 21},
            {name: '18 #2', age: 18},
            {name: '17 #2', age: 17},
            {name: '21 #3', age: 21},
            {name: '19 #1', age: 19},
        ],
        rearranged: [
            {name: '17 #1', age: 17},
            {name: '17 #2', age: 17},
            {name: '18 #1', age: 18},
            {name: '18 #2', age: 18},
            {name: '19 #1', age: 19},
            {name: '21 #1', age: 21},
            {name: '21 #2', age: 21},
            {name: '21 #3', age: 21},
        ]
    },
];

for (const test of tests) {
    const actual = test.students.slice();
    rearrangeInPlace(actual);
    assert.deepStrictEqual(actual, test.rearranged, `For students ${inspect(test.students)}, ` +
        `expected rearranged to be ${inspect(test.rearranged)}, but was ${inspect(actual)}`);
}
